### PR TITLE
Add Hyperopt integration

### PR DESCRIPTION
**Pull request by yuhong-sigopt**

### CODE REVIEW

File: integration_test/test_hyperopt.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import sigopt
+import numpy as np
+import pytest
+
+def objective(p, threshold=0.8):
+  x, y = p['x'], p['y']
+  w = np.random.rand()
+  if w <= threshold:
+    return {
+    'loss': x ** 2 + y ** 2,
```
Comment (gustavomalkomes) [inline integration_test/test_hyperopt.py:13]: lint here

File: integration_test/test_hyperopt.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
+from hyperopt import fmin, tpe, hp, STATUS_OK, STATUS_FAIL, Trials
+import hyperopt
+from sigopt.hyperopt import SigOptTrials, upload_trials
+import sigopt
+import numpy as np
+import pytest
+
+def objective(p, threshold=0.8):
```
Comment (gustavomalkomes) [inline integration_test/test_hyperopt.py:8]: let's have a test with `successful_objective` and one that crashes

File: integration_test/test_hyperopt.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    trials = SigOptTrials(project=project, online=(online and upload))
+    try:
+      best = fmin(lambda x: objective(x, threshold),
+                  space={
+                    'x' : hp.uniform('x', -10, 10),
+                    'y': hp.uniform('y', -10, 10)
+                  },
+                  algo=tpe.suggest,
+                  max_evals=max_evals,
+                  trials=trials)
```
Comment (gustavomalkomes) [inline integration_test/test_hyperopt.py:35]: lint here too

File: integration_test/test_hyperopt.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    except hyperopt.exceptions.AllTrialsFailed:
+      best = None
+    if upload and (online == False):
```
Comment (gustavomalkomes) [inline integration_test/test_hyperopt.py:38]: `online is False`

File: integration_test/test_hyperopt.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    if not upload:
+      assert len(trials.uploaded_tids) == 0
+      return
+    assert len(trials.uploaded_tids) == max_evals
+    self._verify_uploaded_trials(trials, trials.uploaded_tids)
+
+  def _verify_uploaded_trials(self, trials, uploaded_tids):
+    trial_dict = {tid: (result, parameters)
+            for tid, result, parameters in zip(trials.tids, trials.results,
+                                               trials.parameters)}
```
Comment (gustavomalkomes) [inline integration_test/test_hyperopt.py:79]: lint

File: sigopt/hyperopt/base.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    self.factory = SigOptFactory(project)
+    self.online = online
+    self._trials = trials if trials else Trials()
+    self.uploaded_tids = {}
+
+  @property
+  def _dynamic_trials(self):
+    return self._trials._dynamic_trials
+
+
```
Comment (gustavomalkomes) [inline sigopt/hyperopt/base.py:18]: extra line

File: sigopt/hyperopt/base.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+      verbose=verbose,
+      max_queue_len=max_queue_len,
+      allow_trials_fmin=False,  # -- prevent recursion
+      pass_expr_memo_ctrl=pass_expr_memo_ctrl,
+      catch_eval_exceptions=catch_eval_exceptions,
+      return_argmin=return_argmin,
+      show_progressbar=show_progressbar,
+      early_stop_fn=early_stop_fn,
+      trials_save_file=trials_save_file,
+    )
```
Comment (gustavomalkomes) [inline sigopt/hyperopt/base.py:219]: I think most of this can just be 

```
  def __getattr__(self, name):
    try:
      return getattr(self._trials, name)
    except AttributeError:
      raise AttributeError(f"{type(self.__name__)} object has no attribute {name}")
```

File: sigopt/hyperopt/base.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+      if status in [STATUS_OK, STATUS_FAIL]:
+        tid = trial['tid']
+        if tid not in self.uploaded_tids:
+          new_trials.append(trial)
+    ids = self._upload(new_trials)
+    self.uploaded_tids.update(ids)
+    return ids
+
+
+  def _upload(self, trials):
```
Comment (gustavomalkomes) [inline sigopt/hyperopt/base.py:38]: I was expecting this to be `_upload_trial` of only one trial. Not sure why you are filtering them and then adding them. 

But in any case, we need to make sure that we can create so many runs that fast. For the offline case, this could be problematic, so we might need to split the list somehow or add a delay.

We probably should think about this in the context of the python client's `factory.create_multiple_runs(metadata=metadata, list_of_runs_attributes)` 
where `list_of_runs_attributes` is a list of dicts.
What do you think about that?

**Review by gustavomalkomes — CHANGES_REQUESTED** [review]
It's looking good! Thanks for coding this super fast!
I left some comments to simplify the code and so minors with respect to linting but the main points are:

- we need to think about uploading multiple runs at once. Maybe the python client already deals with that, not sure, but it would be cool to have a function in the python client that uploads multiple runs given a list of dictionaries with run parameters. And we can reuse that function for future integrations.
- We just need to add more tests to catch edge cases like when the `fmin` crashes and or we have async trials. Are we not using `pytest` for unit tests in the client?

File: sigopt/hyperopt/base.py (hunk 8 above)
Comment (yuhong-sigopt) [reply sigopt/hyperopt/base.py:38]: I think batch upload API is more efficient and can be used for implement `_upload_trial`.

File: sigopt/hyperopt/base.py (hunk 7 above)
Comment (yuhong-sigopt) [reply sigopt/hyperopt/base.py:219]: It is good! I will try it.

File: integration_test/test_hyperopt.py (hunk 1 above)
Comment (yuhong-sigopt) [reply integration_test/test_hyperopt.py]: done.

File: integration_test/test_hyperopt.py (hunk 2 above)
Comment (yuhong-sigopt) [reply integration_test/test_hyperopt.py:9]: done.

File: integration_test/test_hyperopt.py (hunk 3 above)
Comment (yuhong-sigopt) [reply integration_test/test_hyperopt.py]: done

File: integration_test/test_hyperopt.py (hunk 4 above)
Comment (yuhong-sigopt) [reply integration_test/test_hyperopt.py]: done

File: integration_test/test_hyperopt.py (hunk 5 above)
Comment (yuhong-sigopt) [reply integration_test/test_hyperopt.py:92]: done

File: sigopt/hyperopt/base.py (hunk 6 above)
Comment (yuhong-sigopt) [reply sigopt/hyperopt/base.py]: done

File: sigopt/hyperopt/base.py (hunk 7 above)
Comment (yuhong-sigopt) [reply sigopt/hyperopt/base.py:216]: done.

File: integration_test/test_hyperopt.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  def test_fmin(self, online, upload, objective, max_evals=3):
+    trials, best = self.run_fmin(online, upload, objective, max_evals)
+    self._verify_best_trial(best, trials, objective)
+    self._verify_runs(trials, upload, max_evals)
+
+  def _verify_best_trial(self, best, trials, objective):
+    if objective == 'fail':
+      assert not best
+    if objective == 'success':
+      assert best
```
Comment (tjs-intel) [inline integration_test/test_hyperopt.py:75]: prefer to use `assert best is None` and `assert best is not None` if appropriate, otherwise `0` could cause the test to fail

File: sigopt/hyperopt/base.py [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    if self.online:
+      self.upload()
+
+  # Wrapped APIs
+
+  def view(self, exp_key=None, refresh=True):
+    return self._tails.view(exp_key, refresh)
+
+  def aname(self, trial, name):
+    return self.aname(trial, name)
```
Comment (tjs-intel) [inline sigopt/hyperopt/base.py:84]: !!! infinite recursion!

File: sigopt/hyperopt/base.py (hunk 7 above)
Comment (tjs-intel) [reply sigopt/hyperopt/base.py:216]: did you forget to push your changes? I don't see it

**Review by tjs-intel — CHANGES_REQUESTED** [review]

**Comment by yuhong-sigopt** [comment]
> * we need to think about uploading multiple runs at once. Maybe the python client already deals with that, not sure, but it would be cool to have a function in the python client that uploads multiple runs given a list of dictionaries with run parameters. And we can reuse that function for future integrations.

The `upload_trials` already supports upload list of trial.  There is a test case covered when all trials fail, `fmin` throws exception.  

It seems `async` trial need use `MongoTrials` or `SparkTrials`, more work is need to be done to setup the test environment.

File: integration_test/test_hyperopt.py (hunk 9 above)
Comment (yuhong-sigopt) [reply integration_test/test_hyperopt.py]: Done.

File: sigopt/hyperopt/base.py (hunk 10 above)
Comment (yuhong-sigopt) [reply sigopt/hyperopt/base.py]: Remove.

File: sigopt/hyperopt/base.py (hunk 7 above)
Comment (yuhong-sigopt) [reply sigopt/hyperopt/base.py]: New code is pushed.

File: sigopt/hyperopt/base.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    return self._trials[item]
+
+  def __getattr__(self, name):
+    attrs = dir(self)
+    if name in attrs:
+      return attrs[name]
+    try:
+      return getattr(self._trials, name)
+    except AttributeError as e:
+      raise AttributeError(f"{type(self.__name__)} object has no attribute {name}") from e
```
Comment (gustavomalkomes) [inline sigopt/hyperopt/base.py:103]: this leads to `RecursionError` and also "type` is a `str`. If you change it to 
`raise AttributeError(f"{self.__class__.__name__} object has no attribute {name}") from e` 
that should do it

File: sigopt/hyperopt/base.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
+from .compat import Trials, STATUS_OK, STATUS_FAIL
+from .. import SigOptFactory
+
+HYPEROPT_SOURCE_NAME = 'Hyperopt Suggest'
+HYPEROPT_SOURCE_PRIORITY = 1
+
+
+class SigOptTrials(object):
+  def __init__(self, project, trials=None, online=True):
+    self.factory = SigOptFactory(project)
```
Comment (gustavomalkomes) [inline sigopt/hyperopt/base.py:10]: can we make sure that `project` here has the correct type? 
In particular, I'm thinking more about a nice error message if they try to do: `trials = SigOptTrials(trials)`

Comment (yuhong-sigopt) [reply sigopt/hyperopt/base.py:10]: `SigOptFactory(project)` will check `project` is valid or not.

File: sigopt/hyperopt/base.py (hunk 11 above)
Comment (yuhong-sigopt) [reply sigopt/hyperopt/base.py]: Changed.

File: sigopt/hyperopt/base.py (hunk 12 above)
Comment (gustavomalkomes) [reply sigopt/hyperopt/base.py:13]: oh, I know that, but my point is that a typical Hyperopt user could be confused about the current error; it would be nice to have a helpful error message that says we expect a `project` name for organizing your Trials.
